src/Microsoft.AspNet.Mvc.Core/TempDataDictionary.cs

```diff
@@ -0,0 +1,272 @@
+            _initialKeys.Add(key);
+        }
+
+        public void Clear()
+        {
+            Load();
+            _data.Clear();
+            _retainedKeys.Clear();
+            _initialKeys.Clear();
+        }
+
+        public bool ContainsKey(string key)
+        {
+            Load();
+            return _data.ContainsKey(key);
+        }
+
+        public bool ContainsValue(object value)
+        {
+            Load();
+            return _data.ContainsValue(value);
+        }
+
+        public IEnumerator<KeyValuePair<string, object>> GetEnumerator()
+        {
+            Load();
+            return new TempDataDictionaryEnumerator(this);
+        }
+
+        public bool Remove(string key)
+        {
+            Load();
+            _retainedKeys.Remove(key);
+            _initialKeys.Remove(key);
+            return _data.Remove(key);
+        }
+
+        public bool TryGetValue(string key, out object value)
+        {
+            Load();
+            _initialKeys.Remove(key);
+            return _data.TryGetValue(key, out value);
+        }
+
+        void ICollection<KeyValuePair<string, object>>.CopyTo(KeyValuePair<string, object>[] array, int index)
+        {
+            Load();
+            ((ICollection<KeyValuePair<string, object>>)_data).CopyTo(array, index);
+        }
+
+        void ICollection<KeyValuePair<string, object>>.Add(KeyValuePair<string, object> keyValuePair)
+        {
+            Load();
+            _initialKeys.Add(keyValuePair.Key);
+            ((ICollection<KeyValuePair<string, object>>)_data).Add(keyValuePair);
+        }
+
+        bool ICollection<KeyValuePair<string, object>>.Contains(KeyValuePair<string, object> keyValuePair)
+        {
+            Load();
+            return ((ICollection<KeyValuePair<string, object>>)_data).Contains(keyValuePair);
+        }
+
+        bool ICollection<KeyValuePair<string, object>>.Remove(KeyValuePair<string, object> keyValuePair)
+        {
+            Load();
+            _initialKeys.Remove(keyValuePair.Key);
+            return ((ICollection<KeyValuePair<string, object>>)_data).Remove(keyValuePair);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            Load();
+            return new TempDataDictionaryEnumerator(this);
+        }
+
+        private sealed class TempDataDictionaryEnumerator : IEnumerator<KeyValuePair<string, object>>
+        {
+            private IEnumerator<KeyValuePair<string, object>> _enumerator;
+            private TempDataDictionary _tempData;
+
+            public TempDataDictionaryEnumerator(TempDataDictionary tempData)
+            {
+                _tempData = tempData;
+                _enumerator = _tempData._data.GetEnumerator();
+            }
+
+            public KeyValuePair<string, object> Current
+            {
+                get
+                {
+                    var kvp = _enumerator.Current;
+                    _tempData._initialKeys.Remove(kvp.Key);
```

Comment:
Can you explain this line? I think I understand it but I want to confirm why we're removing the key.